### PR TITLE
Faster ModSummary fingerprints

### PR DESCRIPTION
**Pull request by pepeiborra**
The computation of the GetModSummary rule fingerprints was very suboptimal.
This change:

- Avoids calling Hdr.getOptions twice
- Shares the relevant part of the fingerprint between GetModSummary and GetModSummaryWihoutTimestamps
- skips the timestamps altogether, since we already fingerprint the preprocessed buffer.
- Avoids calling show in the fingerprint computation
- Uses efficient fingerprint primitives

### CODE REVIEW

File: ghcide/src/Development/IDE/Core/Compile.hs [hunk 1]
```diff
@@ -749,7 +752,42 @@ getModSummaryFromImports env fp modTime contents = do
...
+                    , b <- toBytes $ uniq p
+                    ]
+            fingerPrintImports <- withArrayLen moduleUniques $ \len p ->
+                    fingerprintData p len
+            return $ fingerprintFingerprints $
+                    [ fingerprintString fp
+                    , fingerPrintImports
+                    ] ++ map fingerprintString opts
+
+        toBytes :: Int -> [Word8]
```
Comment (wz1000) [inline ghcide/src/Development/IDE/Core/Compile.hs:779]: Is there a way to avoid assuming 64-bit ints here?

Comment (pepeiborra) [reply ghcide/src/Development/IDE/Core/Compile.hs:779]: Good point, thanks

File: ghcide/src/Development/IDE/Core/Rules.hs [hunk 2]
```diff
@@ -873,47 +871,30 @@ getModSummaryRule :: Rules ()
...
-                let ms = msWithTimestamps {
+            Just res@ModSummaryResult{..} -> do
+                let ms = msrModSummary {
                     ms_hs_date = error "use GetModSummary instead of GetModSummaryWithoutTimestamps",
                     ms_hspp_buf = error "use GetModSummary instead of GetModSummaryWithoutTimestamps"
                     }
-                dflags <- hsc_dflags . hscEnv <$> use_ GhcSession f
-                let fp = BS.pack $ show $ hash (computeFingerprint f (fromJust $ ms_hspp_buf msWithTimestamps) dflags ms)
-                return (Just fp, ([], Just res))
+                    fp = fingerprintToBS msrFingerprint
```
Comment (wz1000) [inline ghcide/src/Development/IDE/Core/Rules.hs:895]: Do we not include the buffer in the fingerprint anymore?

Comment (pepeiborra) [reply ghcide/src/Development/IDE/Core/Rules.hs:895]: This was a bug. 
The buffer is included in the `GetModSummary` fingerprint, but should have never been included in the `GetModSummaryWithoutTimestamps` fingerprint, which doesn't include the buffer.

Comment (wz1000) [reply ghcide/src/Development/IDE/Core/Rules.hs:895]: The deleted line is:

```
                 let fp = BS.pack $ show $ hash (computeFingerprint f (fromJust $ ms_hspp_buf msWithTimestamps) dflags ms)
```

Seems like the buffer was included in the fingerprint?

Comment (pepeiborra) [reply ghcide/src/Development/IDE/Core/Rules.hs:895]: That was a bug, because in line 893 the buffer was elided

**Review by wz1000 — APPROVED** [review]

**Pull request merged**